### PR TITLE
refactor(cli): replace RNC cli server api with own Metro middleware

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This drops our usage of `@react-native-community/cli-server-api` in favor of an updated and stripped version of the Metro middleware. It's required for React Native 0.76, as RNC CLI isn't installed through React Native anymore.

## Follow-up

There a few places where we undo some of the features of the previous `@react-native-community/cli-server-api` implementation. This PR only adds a bare-bones version of this package that enables the basics of React Native, without supporting the legacy endpoints.

We'd still need to do a clean up to remove things like:
 
 - https://github.com/expo/expo/blob/e2a9ff69f696a7df74285c80f48b90294808562f/packages/%40expo/cli/src/start/server/middleware/createBuiltinAPIRequestHandler.ts#L31-L35

# How

Based on the [original usage](https://github.com/react-native-community/cli/blob/main/packages/cli-server-api/src/index.ts#L38-L73).

**Server endpoints**

- [X] ~~`/debugger-ui`~~ → _Dropped legacy endpoint. This is now handled by `@react-native/dev-middleware` itself, using the new Chrome DevTools fork._
- [X] ~~`/launch-js-devtools`~~ → _Dropped legacy endpoint. This is now handled by `@react-native/dev-middleware` and our own `/_expo/debugger` handlers._
- [x] `/open-stack-frame` → _This is ported over_
- [x] ~~`/open-url`~~ → _Dropped legacy endpoint. Unclear where this is used, CLI uses app schemes to open the app._
- [X] `/status` → _This is ported over. Also fixed the `process.cwd` as project root header, this is now using the actual loaded project root instead._
- [x] `/symbolicate` → Added `rawBodyMiddleware` to support Metro loading symbolication requests, [see Metro](https://github.com/facebook/metro/blob/main/packages/metro/src/Server.js#L1265-L1266).
- [x] ~~`/systrace`~~ → _Dropped legacy endpoint. This is now handled by `@react-native/dev-middleware`._
- [x] ~~`/reload`~~ → _Dropped legacy endpoint. Unclear where this is used, CLI uses the reload command over websockets._

**Server middlewares**

- [x] `securityHeadersMiddleware` → Fully replaced by Expo's CORS middleware, to support Expo Router features.
- [x] `compression` → Unclear why we would need to compress when running a local/network server. -- **EDIT** Due to bundles being large, compression actually makes bundle loads faster, re-added this in #31732
- [x] `nocache` → This is ported over
- [ ] `serveStatic` (watchFolders) → Unclear why we would need to statically serve the `watchFolder` contents. Both the bundle and asset endpoints would handle this.
- [x] ~~`errorhandler`~~ → Neither Expo or Metro middleware are using `next(new Error)`. Because of that, we can skip this middleware.

**Websocket endpoints**

- [x] ~~`/debugger-proxy`~~ → _Dropped legacy endpoint. This is now handled by `@react-native/dev-middleware` itself, using the new Chrome DevTools fork._
- [x] `/message` → This is ported over
- [x] `/events` → This is ported over

# Test Plan

This PR replaces a core library, because of that, this PR has a big impact on the code base. Some examples that this affects:
- The ability of the native code to determine if Metro is running (through `GET /status`)
- The interactive commands executed from the terminal, like reload of open dev menu (through websockets)
- The ability to debug through the React Native DevTools (through websockets and endpoints from `@react-native/dev-middleware`)

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/server/middleware/suppressErrorMiddleware.ts [hunk 1]
```diff
@@ -1,17 +0,0 @@
-import type { IncomingMessage, ServerResponse } from 'http';
```
Comment (gabrieldonadel) [inline packages/@expo/cli/src/start/server/middleware/suppressErrorMiddleware.ts:1]: Nice!

**Review by gabrieldonadel — APPROVED** [review]
LGTM

File: packages/@expo/cli/src/start/server/middleware/suppressErrorMiddleware.ts (hunk 1 above)
Comment (byCedric) [reply packages/@expo/cli/src/start/server/middleware/suppressErrorMiddleware.ts:1]: We don't need to suppress what doesn't exists 😄

**Comment by byCedric** [comment]
@Simek did you change anything related to eslint by any chance? Its giving linting issues for code this PR doesn't touch. As far as I can see, it could be related to https://github.com/expo/expo/pull/31571

**Pull request merged**